### PR TITLE
#467 Zero Inflated Poisson Distribution

### PR DESCRIPTION
**Pull request by kaijennissen**
#467

Added Zero Inflated Poisson Distribution

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by kashif** [comment]
@kaijennissen you can use my zero inflated class and port that over, then this general class will allow for ZIP as well as ZINB and other distributions. What do you think? https://github.com/zalandoresearch/pytorch-ts/blob/master/pts/distributions/zero_inflated.py

**Comment by PascalIversen** [comment]
Hey @kaijennissen, thanks for looking into this! I think we could also implement the ZIP it like this in the poisson.py:
```python
class ZeroInflatedPoissonOutput(MixtureDistributionOutput):
    def __init__(self):
        super().__init__([PoissonOutput(), DeterministicOutput(0)])
``` 
I would say defining the distribution explicitly is not necessarily needed as it can be accessed with `ZeroInflatedPoissonOutput().distribution()`. Keeping the test you wrote would still be great. What do you think of this approach?

**Comment by kaijennissen** [comment]
@PascalIversen I'll move the code to `poisson.py` and and also inherit from `MixtureDistributionOutput`.
But I have a question regarding the use of `ZeroInflatedPoissonOutput().distribution()`. Does that mean we do not need a `ZeroInflatedPoisson` class? Or just wrap the class around `ZeroInflatedPoissonOutput().distribution()`?

**Comment by PascalIversen** [comment]
Thanks @kaijennissen. Yes I think we would not need it. The main purpose of the Distribution classes is to define how sampling and the   log likelihood works, and that's implicitly defined trough the MixtureDistribution. So I'd say it is fine to just define the Output class.

**Comment by lostella** [comment]
But it looks like @kaijennissen explicitly redefined the `log_prob` method, apart from extending `MixtureDistribution`: my understanding is that this is needed for the same reasons for which we have this in the case of [inflated beta](https://github.com/awslabs/gluon-ts/blob/13eb1f5396240232ddc6ea04ad37eb8fbf3af150/src/gluonts/mx/distribution/inflated_beta.py#L78)? If so, then the `DistributionOutput` should use `ZeroInflatedPoisson`, right @PascalIversen?

I think what @kashif is proposing is to tackle all zero inflated cases at once, given that there's so much in common. Then it probably makes sense to consolidate Poisson, Beta, NB.

**Comment by PascalIversen** [comment]
The inflated beta distribution is special because the beta is not defined on {0,1}. I don't think the zero inflated poisson will result in any nan gradients, since the poisson distribution supports zeros. (I think I checked that once, I guess it makes sense to also add a test for the gradient at the edge case x=0)

**Comment by PascalIversen** [comment]
No nan gradients:

```python
from gluonts.mx.distribution.mixture import MixtureDistributionOutput
from gluonts.mx.distribution.deterministic import DeterministicOutput
from gluonts.mx.distribution.poisson import PoissonOutput
import mxnet as mx

class ZeroInflatedPoissonOutput(MixtureDistributionOutput):
    def __init__(self):
        super().__init__([PoissonOutput(), DeterministicOutput(0)])
        
zip_distr = ZeroInflatedPoissonOutput().distribution(distr_args=[mx.nd.array([[0.4, 0.6]]), mx.nd.array([3.]), mx.nd.array([0.])])        

inputs = mx.nd.array([0., 1.])
inputs.attach_grad()
with mx.autograd.record():
    result = zip_distr.log_prob(inputs)
result.backward()

print(result)
inputs.grad

[-0.4781732 -2.8176785]
<NDArray 2 @cpu(0)>

[0.05383615 0.675828  ]
<NDArray 2 @cpu(0)>
```



Maybe instead of making a superclass for inflated distributions we could think about changing the `MixtureDistribution`'s log_prob to account for mixtures of distributions with non-overlapping support, like the inflated beta. The problem boils down to:

```python
import mxnet as mx
inputs = mx.nd.array([0., 0.5])
inputs.attach_grad()
with mx.autograd.record():
    result = mx.nd.log(inputs)
    result = mx.nd.where(inputs != 0., result, result.zeros_like())
result.backward()
inputs.grad

[nan  2.]
<NDArray 2 @cpu(0)>
```

I wondered if that is a bug, but torch exhibits the same behaviour:
```python
import torch
inputs = torch.tensor([0., 0.5], requires_grad=True)
result = torch.log(inputs)
result =torch.where(inputs != 0., result, torch.tensor([0., 0.]))
result = result.mean()
result.backward()
inputs.grad

tensor([nan, 1.])
```
but in torch one could use an if else statement to circumvent this.

**Comment by kaijennissen** [comment]
Then I'll also add the inflated NB using the solution suggested by @PascalIversen. For the `InflatedGamma` I would also expect `inf`'s / `nan`'s and therefore use the solution of the `ÌnflatedBeta`.

**Comment by kaijennissen** [comment]
@PascalIversen Wouldn't we expect the gradient of the ZIP to be zero for `x=0`? Since x is not part of the pmf for the zero case.
https://en.wikipedia.org/wiki/Zero-inflated_model

**Comment by PascalIversen** [comment]
x is part of the PMF, it just simplifies since 0!=1. 
Thats the gradient we are searching:
image

which is 
image.

**Comment by kaijennissen** [comment]
Of couse your're right. I just looked at the simplified equation. Than it's good that we talked about this, since the gradient produced by my method is zero (Possibly due to the use of  `Deterministic(0)` inside the mixture?).

**Comment by PascalIversen** [comment]
Interesting, I think it is not because of the Deterministic(0) (I am using that too), but in your implementation changing:
```python
# log_prob of zeros
        zero_log_like = self.zero_probability + (
            1 - self.zero_probability
        ) * F.exp(-self.rate) 
```
to
```python
# log_prob of zeros
        zero_log_like = self.zero_probability + (
            1 - self.zero_probability
        ) * self.poisson_distribution.log_prob(inputs)
```
would yield the same log_prob but the correct gradient :-D.

I am wondering why the maximum likelihood parameter inference test you implemented worked anyway....

File: test/distribution/test_mx_distribution_inference.py [hunk 1]
```diff
@@ -1144,3 +1150,103 @@ def test_genpareto_likelihood(xi: float, beta: float, hybridize: bool) -> None:
...
+@pytest.mark.parametrize("hybridize", [False, True])
+def test_inflated_neg_binomial_likelihood(
+    mu_alpha: float,
+    zero_probability: float,
+    hybridize: bool,
+) -> None:
+    """
+    Test to check that maximizing the likelihood recovers the parameters
+    """
+    mu, alpha = mu_alpha
```
[the remark this answers is not in the document]
Comment (kaijennissen) [reply test/distribution/test_mx_distribution_inference.py]: fixed.

File: test/distribution/test_distribution_sampling.py [hunk 2]
```diff
@@ -299,3 +300,109 @@ def test_inflated_beta_sampling(
...
+    )
+
+
+# @pytest.mark.parametrize("mu_alpha", [(5.0, 0.5)])
+# @pytest.mark.parametrize("zero_probability", [0.0, 0.3, 0.9])
+# def test_inflated_neg_binomial_sampling(
+#     mu_alpha: float,
+#     zero_probability: float,
+# ):
+#     mu, alpha = mu_alpha
```
Comment (lostella) [inline test/distribution/test_distribution_sampling.py:363]: @kaijennissen is this test meant to run? If not, it's maybe better to decomment it and `@pytest.mark.skip(...)`-it

Other than that, the PR looks good to me :+1:

Comment (kaijennissen) [reply test/distribution/test_distribution_sampling.py:363]: @lostella I was not sure whether this test is necessary or not since the `ZeroInflatedNegativeBinomialOutput` is a subclass of `MixtureDistributionOutput` which has its own tests. What do you think?

Comment (lostella) [reply test/distribution/test_distribution_sampling.py:363]: I think it's fine not to have it, but then maybe you could just remove the commented lines (just to avoid confusion, no big deal)

**Comment by PascalIversen** [comment]
The `test_inflated_neg_binomial_likelihood` seems to fail on windows. I actually experienced before that maximum likelihood inference with an inflated negative binomial is quite delicate. Could be that the optimization gets stuck in some local minimum. 
@kaijennissen, could you try using more epochs for the maximum likelihood inference test of the inflated negative binomial?

If this does not help we could think about explicitly testing the log_prob() and its gradients instead.

File: test/distribution/test_distribution_sampling.py (hunk 2 above)
Comment (kaijennissen) [reply test/distribution/test_distribution_sampling.py]: Removed.

**Comment by kashif** [comment]
@PascalIversen indeed the ZIB and ZINB type distributions when parametrised by probs are a bit unstable in my experience... on the pytorch side i did it via logits (apart from the gate) but there is fully logit based parametrization now in pyro which could be used but my current version using logits for Poisson or NB is working without issue...

**Comment by PascalIversen** [comment]
Thanks @kashif, interesting point. We could think about basing the NB and maybe Poisson, Binomial on logits. This could stabelize the ZINB inference. (Out of scope of this PR though I think)

File: test/distribution/test_mx_distribution_inference.py [hunk 3]
```diff
@@ -150,6 +153,8 @@ def maximum_likelihood_estimate_sgd(
                 if not hybridize:
                     assert loss.shape == distr.batch_shape
             loss.backward()
+            # import pdb
+            # pdb.set_trace()
```
Comment (PascalIversen) [inline test/distribution/test_mx_distribution_inference.py:157]: This change was probably not intended? I think after removing this we can merge.

**Review by PascalIversen — CHANGES_REQUESTED** [review]

File: test/distribution/test_mx_distribution_inference.py (hunk 3 above)
Comment (kaijennissen) [reply test/distribution/test_mx_distribution_inference.py]: Removed.

**Review by PascalIversen — APPROVED** [review]

**Comment by PascalIversen** [comment]
The tests seem stable now. If that changes or the distributions do not perform well with time series models, we should implement the suggested changes by @kashif for the negative binomial.

Thank you a lot for contributing this @kaijennissen!
